### ISSUE TITLE
beamsearch arguments gets TypeError: coercing to Unicode: need string or buffer, list found

### ISSUE THREAD

**Issue by loretoparisi** [libindic/indic-trans#30]
Hello, I have added `beamsearch` in the arguments to be passed to the like `Transliterator`

```python
# initialize transliterator object
    decode='viterbi'
    if args.beamsearch:
        decode='beamsearch'
    trn = Transliterator(args.source,
                            args.target,
                            rb=args.rb,
                            build_lookup=args.build_lookup,
                            decode=decode)
```

but I get an error when using it:

```
Traceback (most recent call last):
  File "/usr/local/bin/indictrans", line 10, in <module>
    sys.exit(main())
  File "/Library/Python/2.7/site-packages/indictrans/__init__.py", line 138, in main
    process_args(args)
  File "/Library/Python/2.7/site-packages/indictrans/__init__.py", line 129, in process_args
    ofp.write(tline)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/codecs.py", line 357, in write
    data, consumed = self.encode(object, self.errors)
TypeError: coercing to Unicode: need string or buffer, list found
```

**Comment by loretoparisi** [comment]
@irshadbhat thanks, basically I did not change the `Transliterator` code, just added to the `ArgumentsParser` an additional argument to pass the decoding algorithm: `viterbi` (that is by defaults) or `beamsearch`. Do you think then that the output changes when using `beamsearch`?

**Comment by irshadbhat** [comment]
Yes, `beamserach` returns a list instaed of a sigle-best output. `Viterbi` returns the first-best output as per the model parameters, but that output can be incorrect sometimes. We use `beamsearch` to produce n outputs for a source string instead of a single best and later apply some post processing to pick the correct output.

**Comment by loretoparisi** [comment]
@irshadbhat thanks a lot, now it makes sense. Let me try to adapt this to support both, then I will post a PR if you like. Thank you!

**Comment by loretoparisi** [comment]
@irshadbhat I have another question. Do you think that `beamsearch` could help in specific cases rather than `viterbi`? In the paper I read that you are using 2nd order viterbi to decode the best letter sequence, I do not see any mentions (sorry I could be wrong :) of the alternative approach.

Thanks a lot.

**Comment by irshadbhat** [comment]
If we compare first-best of `beamsearch` with `viterbi` search output, `viterbi` will be the winner. `Beamsearch` is usually used for two reasons: **farster** because it only exploits a fixed-length beam (beam-width, in this case 5) instead of an exact search over the whole trellis, secondly, `beamsearch` is used to generate n outputs instead of a single best output. 

Lets suppose we have a noisy English sentence **gd plz sv m** (God please help me). Now assume I model the English normalization task into a transliteration task and the `viterbi` outputs for these words are **good please save my**. If we consider these tokens separately `viterbi` search has actually done a good job but given the context we know **god please save me** is more accurate. This can be achieved using `beamsearch` by generating top n (say 3) transliterations of each word and using some pretrained language model to pick the most likely sentence from the trellis. The trellis would look like:

good          **please**      **save**      my
**god**       pleaze           shave         **me**
guide          plazza           saved          em

I would recommend you to read some good research papers on `beamsearch` and `viterbi search` to get a better idea. You might also find [this one](http://www.aclweb.org/anthology/N18-1090) interesting.
